Quick start for Metal3 (#384)
metal3-io/metal3-docs: docs/user-guide/src/quick-start.md

which.

- bml-01: 80:c1:6e:7a:e8:10
- bml-02: 80:c1:6e:7a:5a:a8

## Virtualized configuration

TODO:

- Create VMs
- Setup sushy-tools

## Common setup

This section is common for both the baremetal configuration and the virtualized
environment. We will go through how to configure and deploy Ironic and Baremetal
Operator.

### Management cluster

If you already have a Kubernetes cluster that you want to use, go ahead and use
that. Please ensure that it is connected to the relevant networks so that Ironic
can reach the BMCs and so that the BareMetalHosts can reach Ironic.

If you do not have an cluster already, you can create one using kind. Please
note that this is absolutely not intended for production environments.

We will use the following configuration file for kind, save it as `kind.yaml`:

```yaml
kind: Cluster
apiVersion: kind.x-k8s.io/v1alpha4
nodes:
  - role: control-plane
    # Open ports for Ironic
    extraPortMappings:
      # Ironic httpd
      - containerPort: 6180
        hostPort: 6180
        listenAddress: "0.0.0.0"
        protocol: TCP
      # Ironic API
      - containerPort: 6385
        hostPort: 6385
        listenAddress: "0.0.0.0"
        protocol: TCP

>       # Inspector API

Reviewer: Is it possible to already remove inspector from here?

      - containerPort: 5050
        hostPort: 5050
        listenAddress: "0.0.0.0"
        protocol: TCP
```

As you can see, it has a few ports forwarded from the host. This is to make
Ironic reachable when it is running inside the kind cluster.

Now go ahead and create the cluster:

```bash
kind create cluster --config kind.yaml
```

We will need to install cert-manager also. It will be used to manage the
certificates for Ironic later.

```bash
kubectl apply -f https://github.com/cert-manager/cert-manager/releases/download/v1.14.3/cert-manager.yaml
```

### DHCP server

The BareMetalHosts must be able to call back to Ironic when going through the
inspection phase. This means that they must have IP addresses in a network where
they can reach Ironic. We will set up a DHCP server for this purpose.

> Any DHCP server can be used for this. We will here use the ironic container

Reviewer: nit: let's be consistent in "Ironic" vs "ironic" (I prefer the former)

image that incudes dnsmasq and some scripts for configuring it.

Create a configuration file like this and save it as `dnsmasq.env`:

```bash
# Specify the MAC addresses (separated by ;) of the hosts we know about and want to use
DHCP_HOSTS=80:c1:6e:7a:e8:10;80:c1:6e:7a:5a:a8
# Ignore unknown hosts so we don't accidentally give out IP addresses to other hosts in the network
DHCP_IGNORE=tag:!known
# Listen on this IP (management computer)
PROVISIONING_IP=192.168.0.150
# Give out IP addresses in this range
DHCP_RANGE=192.168.0.100,192.168.0.149
```

You can now run the DHCP server like this:

```bash

> docker run --name dnsmasq --rm -d --net=host --privileged --user 997:994 \

Reviewer: Bonus for using capabilities instead of --privileged

Author: If you can tell me exactly which to use I'll happily change it :sweat_smile:

  --env-file dnsmasq.env --entrypoint /bin/rundnsmasq \
  quay.io/metal3-io/ironic
```

### Image server

In order to do anything useful, we will need a server for hosting disk images
that can be used to provision the servers.

Create a folder to hold the disk images:

```bash
mkdir disk-images
```

> Download an image to use for testing:

Reviewer: This is supposed to be a real baremetal user guide, right? Let's avoid cirros then. I'm impressed you actually found some hardware where it works, but on most it probably won't (missing drivers for NICs, RAID controllers, etc).

Author: Thanks for the heads up! I was not aware of this

```bash
wget -P disk-images https://github.com/cirros-dev/cirros/releases/download/0.6.2/cirros-0.6.2-x86_64-disk.img
```

Run a basic http server to expose the disk images:

```bash
docker run --name image-server --rm -d -p 80:8080 \
  -v "$(pwd)/disk-images:/usr/share/nginx/html" nginxinc/nginx-unprivileged
```

### Deploy Ironic

In this section we will create a
[kustomization](https://kubectl.docs.kubernetes.io/references/kustomize/glossary/#kustomization)
containing configuration and credentials for deploying Ironic.

Create a folder to hold the kustomization:

```bash
mkdir ironic
```

#### Authentication configuration

Create authentication configuration for Ironic and Inspector. You will need to
generate a username and password for each. We will here refer to them as
`IRONIC_USERNAME`, `IRONIC_PASSWORD`, `INSPECTOR_USERNAME` and
`INSPECTOR_PASSWORD`.

> Create a file `ironic-auth-config` with configuration for how to access Ironic.

Reviewer: This all will be redundant without inspector. Since this guide is still a draft, maybe avoid inspector already? You can set USE_IRONIC_INSPECTOR=false for now.

Author: Unfortunately it turned out to be more work removing inspector everywhere then just using it as is. All the manifests still include and expect Inspector to be there in v0.5.0. I will happily update this once we have a new release that doesn't use Inspector though.
(I hope we can get this guide merged in time for KubeCon :crossed_fingers: )

access Inspector. This will be used by Ironic. It should have the following
content:

```conf
[inspector]
auth_type=http_basic
username=INSPECTOR_USERNAME
password=INSPECTOR_PASSWORD
```

To enable basic auth, we need to create secrets containing the keys
`IRONIC_HTPASSWD` and `INSPECTOR_HTPASSWD` with values generated from the
credentials using htpasswd. We will do this by creating two files
`ironic-htpasswd` and `ironic-inspector-htpasswd` with the following content.

`ironic-htpasswd`:

```bash
IRONIC_HTPASSWD="output of `htpasswd -n -b -B IRONIC_USERNAME IRONIC_PASSWORD`"
```

Similarly for `ironic-inspector-htpasswd`:

```bash
IRONIC_HTPASSWD="output of `htpasswd -n -b -B INSPECTOR_USERNAME INSPECTOR_PASSWORD`"
```

#### Ironic environment variables

In this section we will create a file containing environment variables used to
configure Ironic and related components. We will call the file `ironic_bmo.env`.
It looks like this for the baremetal lab:

```bash
# Same port as exposed in kind.yaml
HTTP_PORT=6180
# This is the interface inside the container
PROVISIONING_INTERFACE=eth0
# URL where the http server is exposed (IP of management computer)
CACHEURL=http://192.168.0.150

> IRONIC_FAST_TRACK=true

Reviewer: Not needed, please remove (I want to deprecate the explicit variable).

   ```

5. Deploy Ironic.

   ```bash
   kubectl apply -k ironic
   ```

6. Deploy Baremetal Operator.

   ```bash
   kubectl apply -k bmo
   ```

## Create BareMetalHosts

Now that we have Baremetal operator deployed, let's put it to use by creating
BareMetalHosts (BMHs) to represent our servers. You will need the protocol and
IPs of the BMCs, as well as credentials for accessing them, and the servers MAC
addresses.

Create one secret for each BareMetalHost, containing the credentials for
accessing its BMC. Here is an example:

```yaml
apiVersion: v1
kind: Secret
metadata:
  name: bml-01
type: Opaque
stringData:
  username: replaceme
  password: replaceme
```

Then continue by creating the BareMetalHost manifest. You can put it in the same
file as the secret if you want. Just remember to separate the two resources with
one line containing `---`.

Here is an example of a BareMetalHost referencing the secret above with MAC
address and BMC address matching our `bml-01` server.

```yaml
apiVersion: metal3.io/v1alpha1
kind: BareMetalHost
metadata:
  name: bml-01
spec:
  online: true
  bootMACAddress: 80:c1:6e:7a:e8:10

>   bootMode: legacy

Reviewer: Is it really required in your case? We no longer recommend using legacy boot by default. At the very least, please explain this choice so that the users don't blindly copy it on newer hardware.

Author: Sadly it is :slightly_frowning_face: 
I actually tried first with UEFI but it was not supported on the hardware. I will add a comment about it though

# Quick-start for Metal3

This guide has been tested on Ubuntu server 22.04. It should be seen as an
example rather than the absolute truth about how to deploy and use Metal3. We
will cover two environments and two scenarios. The environments are

1. a baremetal lab with actual physical servers and baseboard management
   controllers (BMCs), and

> 2. a virtualized baremetal lab with virtual machines and sushy-tools acting as

Reviewer 2: ```suggestion
1. a virtualized baremetal lab with virtual machines and sushy-tools acting as
```
Our generator supports auto numbered lists, right?

If so, I'd like to see it utilized for all numbered lists here.

   BMC.

In both of these, we will show how to use Baremetal operator and Ironic to
manage the servers through a Kubernetes API, as well as how to turn the servers
into Kubernetes clusters managed through Cluster API. These are the two
scenarios.

In a nut-shell, this is what we will do:

> 1. Setup a management cluster

Reviewer 2: Add anchor links to these as well.

2. Setup a DHCP server
3. Setup a disk image server
4. Deploy Ironic
5. Deploy Baremetal operator
6. Create BareMetalHosts to represent the servers
7. (Scenario 1) Provision the BareMetalHosts
8. (Scenario 2) Deploy Cluster API and turn the BareMetalHosts into a Kubernetes
   cluster

## Prerequisites

You will need the following software installed.

> - docker (or podman)

Reviewer 2: Docker as a softwarere is a name, hence it would need to be capitalized. docker as a CLI command would not. So either fix the capitalization of all the tools, or alternatively L31 could read tools instead of software?

image that incudes dnsmasq and some scripts for configuring it.

Create a configuration file and save it as `dnsmasq.env`.

Baremetal lab:

```bash
# Specify the MAC addresses (separated by ;) of the hosts we know about and want to use
DHCP_HOSTS=80:c1:6e:7a:e8:10;80:c1:6e:7a:5a:a8
# Ignore unknown hosts so we don't accidentally give out IP addresses to other hosts in the network
DHCP_IGNORE=tag:!known
# Listen on this IP (management computer)
PROVISIONING_IP=192.168.0.150
# Give out IP addresses in this range
DHCP_RANGE=192.168.0.100,192.168.0.149
GATEWAY_IP=192.168.0.1
```

Virtualized environment:

```bash
DHCP_HOSTS=00:60:2f:31:81:01
DHCP_IGNORE=tag:!known
# IP of the host from VM perspective
PROVISIONING_IP=192.168.222.1
GATEWAY_IP=192.168.222.1
DHCP_RANGE=192.168.222.100,192.168.222.149
```

You can now run the DHCP server like this:

```bash
docker run --name dnsmasq --rm -d --net=host --privileged --user 997:994 \
  --env-file dnsmasq.env --entrypoint /bin/rundnsmasq \
  quay.io/metal3-io/ironic
```

### Image server

In order to do anything useful, we will need a server for hosting disk images
that can be used to provision the servers.

> Create a folder to hold the disk images:

Reviewer 2: ```suggestion
Create a directory to hold the disk images:
```
Folder is GUI terminology.

Create a folder to hold the kustomization:

```bash
mkdir ironic
```

#### Authentication configuration

Create authentication configuration for Ironic and Inspector. You will need to
generate a username and password for each. We will here refer to them as
`IRONIC_USERNAME`, `IRONIC_PASSWORD`, `INSPECTOR_USERNAME` and
`INSPECTOR_PASSWORD`.

Create a file `ironic-auth-config` with configuration for how to access Ironic.
This will be use by Inspector. It should have the following content:

```conf
[ironic]
auth_type=http_basic
username=IRONIC_USERNAME
password=IRONIC_PASSWORD
```

Create a file `ironic-inspector-auth-config` with configuration for how to
access Inspector. This will be used by Ironic. It should have the following
content:

```conf
[inspector]
auth_type=http_basic
username=INSPECTOR_USERNAME
password=INSPECTOR_PASSWORD
```

To enable basic auth, we need to create secrets containing the keys
`IRONIC_HTPASSWD` and `INSPECTOR_HTPASSWD` with values generated from the
credentials using htpasswd. We will do this by creating two files
`ironic-htpasswd` and `ironic-inspector-htpasswd` with the following content.

`ironic-htpasswd`:

```bash

> IRONIC_HTPASSWD="output of `htpasswd -n -b -B IRONIC_USERNAME IRONIC_PASSWORD`"

Reviewer 2: ```suggestion
IRONIC_HTPASSWD="<output of `htpasswd -n -b -B IRONIC_USERNAME IRONIC_PASSWORD`>"
```
`<>` commonly used for placeholders/options.

need the dnsmasq container that comes with the kustomization by default.

We will create a patch for removing it. It looks like this:

```yaml
apiVersion: apps/v1
kind: Deployment
metadata:
  name: ironic
spec:
  template:
    spec:
      containers:
        - name: ironic-dnsmasq
          $patch: delete
```

Save it as `ironic-patch.yaml`.

#### Ironic kustomization

Time to tie it all together by creating a `kustomization.yaml`. At this point
you should have a file structure like this:

```text
ironic/
├── ironic-auth-config
├── ironic-htpasswd
├── ironic-inspector-auth-config
├── ironic-inspector-htpasswd
├── ironic-patch.yaml
├── ironic_bmo.env
└── kustomization.yaml
```

Here is a commented `kustomization.yaml`. Check carefully the IP addresses as
these will always differ depending on environment.

```yaml
apiVersion: kustomize.config.k8s.io/v1beta1
kind: Kustomization
namespace: baremetal-operator-system
# These are the kustomizations we build on. You can download them and change the URLs to relative
# paths if you do not want to access them over the network.
# Note that the ref=v0.5.0 specifies the version to use.
resources:

>   - https://github.com/metal3-io/baremetal-operator/config/namespace?ref=v0.5.0

Reviewer 2: Have the YAML in k8s formatting, ie. without leading spaces on the lists.

```

When you are happy with the output, apply it in the cluster:

```bash
kubectl apply -k bmo
```

## Deployment summary

You are not expected to go through all the above steps each time you want to
deploy Metal3. Store the configuration and reuse it the next time.

Here is a summary of the deploy steps when all configuration is already in
place.

1. Create the management cluster.

   ```bash
   kind create cluster --config kind.yaml
   ```

2. Deploy cert-manager.

   ```bash
   kubectl apply -f https://github.com/cert-manager/cert-manager/releases/download/v1.14.3/cert-manager.yaml
   ```

3. Start the DHCP server.

   ```bash
   docker run --name dnsmasq --rm -d --net=host --privileged --user 997:994 \
     --env-file dnsmasq.env --entrypoint /bin/rundnsmasq \
     quay.io/metal3-io/ironic
   ```

4. Start the image server.

   ```bash
   docker run --name image-server --rm -d -p 80:8080 \
     -v "$(pwd)/disk-images:/usr/share/nginx/html" nginxinc/nginx-unprivileged
   ```

5. Deploy Ironic.

   ```bash
   kubectl apply -k ironic
   ```

6. Deploy Baremetal Operator.

   ```bash
   kubectl apply -k bmo
   ```

## Create BareMetalHosts

> Now that we have Baremetal operator deployed, let's put it to use by creating

Reviewer 2: ```suggestion
Now that we have Bare Metal Operator deployed, let's put it to use by creating
```
